### DISCUSSION TITLE
Unspecified beavior in async SQLAlchemy 1.4 when using multiple loader options

### DISCUSSION THREAD

**Discussion by sylencecc** [sqlalchemy/sqlalchemy#8615] [Usage Questions] [answered]
Hello! 
I currently experience an issue with loader options when using async SQLAlchemy 1.14.41 with PostgreSQL. I seem to hit random or undefined behavior and I'm not sure whether that is due to how I specify loader options or if I encountered a bug or race condition in SQLAlchemy itself. I tried to reduce my problematic test case to the following piece of code - further reduction changes behavior for no obvious reasons.

Running the script posted below either succeeds or yields the following exception:
```
Traceback (most recent call last):
  File "/srv/src/sqlalchemy_async_eager_load_issue.py", line 145, in <module>
    asyncio.run(run())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/srv/src/sqlalchemy_async_eager_load_issue.py", line 140, in run
    assert ta.tenant.name == "CERT"
  File "/usr/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 482, in __get__
    return self.impl.get(state, dict_)
  File "/usr/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 942, in get
    value = self._fire_loader_callables(state, key, passive)
  File "/usr/lib/python3.10/site-packages/sqlalchemy/orm/attributes.py", line 978, in _fire_loader_callables
    return self.callable_(state, passive)
  File "/usr/lib/python3.10/site-packages/sqlalchemy/orm/strategies.py", line 856, in _load_for_state
    self._invoke_raise_load(state, passive, "raise")
  File "/usr/lib/python3.10/site-packages/sqlalchemy/orm/strategies.py", line 816, in _invoke_raise_load
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: 'TenantAffiliation.tenant' is not available due to lazy='raise'
```

Which is unexpected since I supplied that specific association as loader option via `selectinload(Tenant.users).selectinload(TenantAffiliation.tenant)`. The weird thing is that the code randomly starts to work when adding or removing seemingly unrelated parts of it. For example, it will (consistently on the system I'm writing this on) run without raising an error after removing the unused `class RolePermission`. On another system the code runs without an exception as-is, but starts to fail (consistently) once `class RolePermission` is removed. My original test case within the actual application ran successfully with `echo=True`, but failed with `echo=False`. I'm sorry for the long convoluted code sample. I tried to shrink it as much as possible, but due to the random nature of this issue, removing more stuff (such as the class RolePermission) suddenly "fixed" the exception without an obvious reason.

```python
import asyncio
from dataclasses import dataclass, field
from typing import Optional, Set
import os

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import registry, relationship, sessionmaker, selectinload
from sqlalchemy.schema import ForeignKey
from sqlalchemy import Column, Integer, String, Table, Boolean
from sqlalchemy.future import select


DB_URI = (
    f'postgresql+asyncpg://{os.getenv("POSTGRES_USER")}'
    f':{os.getenv("POSTGRES_PASSWORD")}'
    f'@{os.getenv("POSTGRES_HOST")}'
    f':{os.getenv("POSTGRES_PORT")}'
    f'/{os.getenv("APP_DB")}'
)


@dataclass(eq=False, kw_only=True)
class User:
    id: Optional[int]
    tenants: Set["TenantAffiliation"] = field(init=False, default_factory=set, repr=False)
    name: str


@dataclass(eq=False, kw_only=True)
class TenantAffiliation():
    role: "Role"
    tenant: "Tenant"
    user: User


@dataclass(eq=False, kw_only=True)
class Tenant:
    id: Optional[int]
    users: Set[TenantAffiliation] = field(init=False, default_factory=set, repr=False)
    name: str


@dataclass(eq=False, kw_only=True)
class RolePermission:
    permission: int
    role: "Role"


@dataclass(eq=False, kw_only=True)
class Role:
    id: Optional[int]
    locked: bool = False
    name: str


async def run():
    orm_registry = registry()
    Table("tenants",
        orm_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("name", String(32), nullable=False)
    )
    Table(
        "users",
        orm_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("name", String(64), nullable=False)
    )
    Table(
        "roles",
        orm_registry.metadata,
        Column("id", Integer, primary_key=True),
        Column("name", String(32), nullable=False),
        Column("locked", Boolean, nullable=False)
    )
    Table(
        "role_permissions",
        orm_registry.metadata,
        Column("permission", Integer, primary_key=True),
        Column("role_id", Integer, ForeignKey("roles.id"), primary_key=True)
    )
    Table(
        "tenant_affiliations",
        orm_registry.metadata,
        Column("tenant_id", Integer, ForeignKey("tenants.id"), primary_key=True),
        Column("user_id", Integer, ForeignKey("users.id"), primary_key=True),
        Column("role_id", Integer, ForeignKey("roles.id"), nullable=False)
    )

    tables = orm_registry.metadata.tables
    orm_registry.map_imperatively(
        Tenant,
        tables["tenants"],
        properties={
            "users": relationship(TenantAffiliation, back_populates="tenant", collection_class=set, lazy="selectin")
        }
    )
    orm_registry.map_imperatively(
        User,
        tables["users"],
        properties={
            "tenants": relationship(TenantAffiliation, back_populates="user", collection_class=set, lazy="raise")
        }
    )
    orm_registry.map_imperatively(
        Role,
        tables["roles"],
    )
    orm_registry.map_imperatively(
        TenantAffiliation,
        tables["tenant_affiliations"],
        properties={
            "tenant": relationship(Tenant, back_populates="users", lazy="raise"),
            "user": relationship(User, back_populates="tenants", lazy="raise"),
            "role": relationship(Role, lazy="raise")
        }
    )

    engine = create_async_engine(DB_URI, echo=False)
    async with engine.begin() as conn:
        await conn.run_sync(orm_registry.metadata.drop_all)
        await conn.run_sync(orm_registry.metadata.create_all)
    session = sessionmaker(bind=engine, class_=AsyncSession, expire_on_commit=False)()
    async with session:
        cert = Tenant(name="CERT", id=None)
        ur = Role(name="User", id=None)
        alice = User(name="Alice", id=None)
        alice_aff = TenantAffiliation(role=ur, tenant=cert, user=alice)
        for entity in [cert, ur, alice, alice_aff]:
            session.add(entity)
            await session.commit()
        qry = select(Tenant).options(
            selectinload(Tenant.users).selectinload(TenantAffiliation.user).selectinload(User.tenants),
            selectinload(Tenant.users).selectinload(TenantAffiliation.tenant))\
            .execution_options(populate_existing=True)
        q_t = (await session.execute(qry)).scalar_one()
        assert isinstance(q_t, Tenant)
        assert q_t.name == "CERT"
        ta = next(iter(q_t.users))
        assert ta.tenant.name == "CERT"
        assert len(ta.user.tenants) == 1


if __name__ == "__main__":
    asyncio.run(run())
```

**Comment by zzzeek** [comment, marked as answer]
the script here itself is complete, but doesnt reproduce the issue for me.   The reason this kind of issue would happen that I'm familiar with would require there to be a more complex set of data in the database for it to occur; basically a particular TenantAffiliation would be first loaded as a result of being at the end of either `.selectinload(User.tenants)` meaning the as the ORM loads rows, that's the first path by which it locates some particular TenantAffiliation.   That TenantAffiliation goes into the identity map with its `.tenant` field unpopulated, and you get the raise.

if TenantAffliation is the important path, you might want to try this:

```py
                selectinload(Tenant.users)
                .selectinload(TenantAffiliation.user)
                .selectinload(User.tenants).joinedload(TenantAffiliation.tenant, innerjoin=True),

```

however, since all these selectinloads basically do an extra SELECT, you could also just load all the TenantAffiliation objects you will need up front instead so that there is no issue w/ pathing.    

this use case is basically cyclical eager loads and this kind of problem is not really avoidable. We have a new feature in 2.0 that will help for extremely simple cases so far called [recursion_depth](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html?highlight=selectinload#sqlalchemy.orm.selectinload.params.recursion_depth) but it doesn't work for multi-entity relationships (the traversal issues are extremely complicated).

**Comment by sylencecc** [comment]
Thanks for the elaborate response!
The script does indeed run as-is on some systems I tested it on, but fails after removing or rearranging some of the code (which shouldn't affect the outcome). The weird behavior seems indeed be connected to the order in which the paths specified in `options()` are followed. Reducing the query to a single path in `options()`, as you suggested, works reliably. However, in case multiple paths are strictly required everything started to work once I set `populate_existing` to `False` (the default). I seem to have misunderstood that parameter and the default (`False`) seems to be what I actually want. That way, I can either (as you suggested) load the TenantAffiliations separately or specify multiple paths in `options()` as in my original script. With `populate_existing=True` the already loaded entities are reloaded based on the current query, which seems to also unload associations not contained in the current query, leading to the original issue (such as depending on which path was evaluated last, either `TenantAffiliation.user` or `TenantAffiliation.tenant` is not loaded).

**Reply by zzzeek** [reply]
hmm I didnt think populate_existing would affect paths that it otherwise would not be loading.  it shouldn't be unconditionally expiring everything. it would just overwrite those attributes which it would be loading for a new object.    that said, when you have cycles in loader options as is the case here, things get very complicated and it's extremely difficult to predict the specifics of the traversal.
